File: challenge_C/level2/쿼드압축 후 개수 세기.c
#include <stdio.h>

//분할정복 알고리즘
int ans[2];

//구역 안의 값이 모두 같은지 확인하는 함수
int check(int row, int col, int size, int **arr)
{
    int start = arr[row][col];
    
    for(int i = row; i < size + row; i++)
        for(int j = col; j < size + col; j++)
            if(start != arr[i][j])
                return 0;
    return 1;
}

//구역을 분할하는 함수
void divide(int row, int col, int size, int **arr)
{
    if(check(row, col, size, arr))
        ans[arr[row][col]]++;
    
    else
    {
        size /= 2;
        
        for(int i = 0; i < 2; i++)
            for(int j = 0; j < 2; j++)
                divide(row + size * i, col + size * j, size, arr);
    }
}

int* solution(int** arr, size_t arr_rows, size_t arr_cols)
{
    divide(0, 0, arr_rows, arr);
    
    return ans;
}

File: challenge_Java/level2/쿼드압축 후 개수 세기.java
class Solution {
    public int[][] standard = {{0, 0}, {0, 1}, {1, 0}, {1, 1}};
    public int[] answer = new int[2];
    
    public int[] solution(int[][] arr) {
        divide(arr, 0, 0, arr.length);
        return answer;
    }
    
    public void divide(int[][] arr, int row, int col, int size) {
        for (int i = 0; i < 4; i++) {
            int nextRow = row + standard[i][0] * size;
            int nextCol = col + standard[i][1] * size;
            
            if (checkNums(arr, nextRow, nextCol, size) == false)
                divide(arr, nextRow, nextCol, size / 2);
            else
                answer[arr[nextRow][nextCol]]++;
            
            if (row == 0 && col == 0 && size == arr.length)
                break;
        }
    }
    
    public boolean checkNums(int[][] arr, int row, int col, int size) {
        int num = arr[row][col];
        
        for (int i = row; i < row + size; i++)
            for (int j = col; j < col + size; j++)
                if (arr[i][j] != num)
                    return false;
        return true;
    }
}
